### PR TITLE
Allow specification of ssh for checkout protocol

### PR DESCRIPTION
**Pull request by matthew-cox**
This is a small PR which allows the user to use ssh for the clone action. This lays the groundwork for private org or team repository checkouts with more secure authentication via ssh key.

### CODE REVIEW

**Comment by anshumanbh** [comment]
HI @matthew-cox, thanks a ton for this PR. I have a couple of questions:

It looks like you have added an extra option to choose whether somebody wants to clone using the `https` URL or the `ssh` URL. Cloning the HTTPS url should work without any need of a key or anything but wouldn't cloning using the SSH protocol require a SSH key? Or, does providing the github user token (of that org/private repo) just work?

Have you tested this?

**Comment by matthew-cox** [comment]
Ok, I've documented private cloning via SSH. I added a basic ssh config for github to the Docker image too. Finally, team support has been added.

I'll note that it would be nice for the documentation if a `latest` tag existed for the published `git-all-secrets` image. Assuming you accept the new ssh changes, the README will need updating with the newest version tag.

Having a `latest` tag would also make it easer on users to get the latest.

Let me know if you have any questions!

Matt

**Comment by matthew-cox** [comment]
@anshumanbh any update on merging this?

**Comment by anshumanbh** [comment]
@matthew-cox sorry havent gotten a chance to look at the code yet. Hopefully, I can get to it in the next few days!

File: main.go [hunk 1]
```diff
@@ -46,6 +49,32 @@ func check(e error) {
...
+
+	// do not clone forks
+	if !*cloneForks && *repo.Fork {
+		return nil
+	} else {
+		// clone it
+		wg.Add(1)
+		fmt.Println(urlToClone)
+		// thread out the git clone
+		go gitclone(urlToClone, directory, wg)
```
Comment (anshumanbh) [inline main.go]: shouldn't this be `&wg` i.e. sending the waitgroup pointer just like others?
Also, shouldn't we be waiting as well by using `wg.Wait()`

File: main.go [hunk 2]
```diff
@@ -76,27 +105,73 @@ func cloneorgrepos(ctx context.Context, client *github.Client, org string) error
 	var orgclone sync.WaitGroup
 	//iterating through the repo array
```
Comment (anshumanbh) [inline main.go]: I think you're missing the `orgclone.Add(1)`?

File: main.go [hunk 3]
```diff
@@ -76,27 +105,73 @@ func cloneorgrepos(ctx context.Context, client *github.Client, org string) error
...
+			check(err)
+			teamRepos = append(teamRepos, repos...) //adding to the repo array
+			if resp.NextPage == 0 {
+				break
 			}
+			listTeamRepoOpts.Page = resp.NextPage
 		}
 
+		var teamclone sync.WaitGroup
+		//iterating through the repo array
```
Comment (anshumanbh) [inline main.go]: `teamclone.Add(1)` ?

**Comment by anshumanbh** [comment]
@matthew-cox I briefly glanced over the new code. I think you might be missing a few things around wait groups. Normally, you define a wait group as a variable, you add(1) to it and pass its pointer to the function. And, then you wait for the waitgroup to return back. 

In the function itself that gets called, you'd call waitgroup.Done. 

Check some of the existing usage of waitgroups in the code and can you please modify your code accordingly?

File: main.go (hunk 1 above)
Comment (matthew-cox) [reply main.go]: `wg` is already a pointer (`*wg` in the function arguments). Passing it without `&` here is passing a pointer.

File: main.go (hunk 2 above)
Comment (matthew-cox) [reply main.go]: `executeclone` takes a pointer to `orgclone` (`& orgclone ` in the call) and makes the `wg.Add()` call for each thread. After calling `executeclone`, `cloneorgrepos` calls `orgclone.Wait()`

File: main.go (hunk 3 above)
Comment (matthew-cox) [reply main.go]: `executeclone` takes a pointer to `teamclone` (`&teamclone` in the call) and makes the `wg.Add()` call for each thread. After calling `executeclone`, `cloneTeamRepos` calls `teamclone.Wait()`

**Comment by matthew-cox** [comment]
@anshumanbh I responded to your comments above. I think that the wait group usage is fine, although it is slightly different than your previous code. Since `executeclone` only clones one repo at a time, I don't think that is where the `Wait()` should be if the intent is to clone repositories in parallel.

**Comment by anshumanbh** [comment]
@matthew-cox Hi Matt, I was finally able to get to this today. I added a bunch of changes to the README along with yours. I also made some minor changes in the code but not anything big. There were some conflicts in the code so I just made changes to my old code and pushed it from there instead of merging your PR. I am going to close this for now but please check out the latest code and the documentation. And, please let me know if you have any issues. Thanks again!!

**Comment by anshumanbh** [comment]
@matthew-cox Hi Matt, after I pushed the code and the Docker image, I tried testing it. It looks like private repositories are not even being listed in the first place. The `cloneuserrepos` function first lists all the repositories before the cloning can begin via the call `client.Repositories.List`. This listing call doesn't list any private repositories of a user. When you tested it, were private repositories being cloned at all using the SSH key?

**Comment by anshumanbh** [comment]
I figured it out. I had to make the user `""` for both listing repos and gists since if you don't specify which user, it will list for the authenticated user. This is so stupid but it took me a while to dig this down from the go-github library. https://github.com/anshumanbh/git-all-secrets/commit/7ae0077d629430004ff1d46a4f95cb6bfe01eace should fix the bug where private repos weren't getting listed. I don't think you need SSH keys anymore. You can simply provide your personal access token and you should be able to scan your private repos as well.

**Comment by matthew-cox** [comment]
Awesome! Thanks your figuring this out. 😄

**Pull request closed without merging**